dubbo-remoting/dubbo-remoting-api/src/main/java/org/apache/dubbo/remoting/exchange/support/header/HeaderExchangeChannel.java

```diff
@@ -1,251 +1,252 @@
+
+    private final Channel channel;
+
+    private volatile boolean closed = false;
+
+    HeaderExchangeChannel(Channel channel) {
+        if (channel == null) {
+            throw new IllegalArgumentException("channel == null");
+        }
+        this.channel = channel;
+    }
+
+    static HeaderExchangeChannel getOrAddChannel(Channel ch) {
+        if (ch == null) {
+            return null;
+        }
+        HeaderExchangeChannel ret = (HeaderExchangeChannel) ch.getAttribute(CHANNEL_KEY);
+        if (ret == null) {
+            ret = new HeaderExchangeChannel(ch);
+            if (ch.isConnected()) {
+                ch.setAttribute(CHANNEL_KEY, ret);
+            }
+        }
+        return ret;
+    }
+
+    static void removeChannelIfDisconnected(Channel ch) {
+        if (ch != null && !ch.isConnected()) {
+            ch.removeAttribute(CHANNEL_KEY);
+        }
+    }
+
+    @Override
+    public void send(Object message) throws RemotingException {
+        send(message, false);
+    }
+
+    @Override
+    public void send(Object message, boolean sent) throws RemotingException {
+        if (closed) {
+            throw new RemotingException(this.getLocalAddress(), null, "Failed to send message " + message + ", cause: The channel " + this + " is closed!");
+        }
+        if (message instanceof Request
+                || message instanceof Response
+                || message instanceof String) {
+            channel.send(message, sent);
+        } else {
+            Request request = new Request();
+            request.setVersion(Version.getProtocolVersion());
+            request.setTwoWay(false);
+            request.setData(message);
+            channel.send(request, sent);
+        }
+    }
+
+    @Override
+    public CompletableFuture<Object> request(Object request) throws RemotingException {
+        return request(request, channel.getUrl().getPositiveParameter(TIMEOUT_KEY, DEFAULT_TIMEOUT));
+    }
+
+    @Override
+    public CompletableFuture<Object> request(Object request, int timeout) throws RemotingException {
+        if (closed) {
+            throw new RemotingException(this.getLocalAddress(), null, "Failed to send request " + request + ", cause: The channel " + this + " is closed!");
+        }
+        // create request.
+        Request req = new Request();
+        req.setVersion(Version.getProtocolVersion());
+        req.setTwoWay(true);
+        req.setData(request);
+        DefaultFuture future = DefaultFuture.newFuture(channel, req, timeout);
+        try {
+            channel.send(req);
+        } catch (RemotingException e) {
+            future.cancel();
+            throw e;
+        }
+        return future;
+    }
+
+    @Override
+    public boolean isClosed() {
+        return closed;
+    }
+
+    @Override
+    public void close() {
+        try {
+            DefaultFuture.closeChannel(channel);
```

Comment:
just modify this line

Reply:
I verified on the master branch. This solution doesn't work.